src/ring_theory/witt_vector/truncated.lean

```diff
@@ -240,3 +238,221 @@ instance : comm_ring (truncated_witt_vector p n R) :=
+by simp only [truncated_witt_vector, fintype.card_fin, fintype.card_fun]
+
+end fintype
+
+lemma infi_ker_truncate : (⨅ i : ℕ, (@witt_vector.truncate p _ i R _).ker) = ⊥ :=
+begin
+  rw [submodule.eq_bot_iff],
+  intros x hx,
+  ext,
+  simp only [witt_vector.mem_ker_truncate, ideal.mem_infi, witt_vector.zero_coeff] at hx ⊢,
+  exact hx _ _ (nat.lt_succ_self _)
+end
+
+end truncated_witt_vector
+
+namespace witt_vector
+open truncated_witt_vector (hiding truncate coeff)
+
+section lift
+
+variable [comm_ring R]
+variables {S : Type*} [comm_ring S]
+variable (f : Π k : ℕ, S →+* truncated_witt_vector p k R)
+variable f_compat : ∀ (k₁ k₂ : ℕ) (hk : k₁ ≤ k₂),
+           (truncated_witt_vector.truncate hk).comp (f k₂) = f k₁
+variables {p R}
+variable (n)
+
+/--
+Given a family `fₖ : S → truncated_witt_vector p k R` and `s : S`, we produce a Witt vector by
+defining the `k`th entry to be the final entry of `fₖ s`.
+-/
+def lift_fun (s : S) : 𝕎 R :=
+witt_vector.mk p $ λ k, truncated_witt_vector.coeff (fin.last k) (f (k+1) s)
+
+variables {f}
+include f_compat
+
+@[simp] lemma truncate_lift_fun (s : S) :
+  witt_vector.truncate n (lift_fun f s) = f n s :=
+begin
+  ext i,
+  simp only [lift_fun, truncated_witt_vector.coeff_mk, witt_vector.truncate_mk],
+  rw [← f_compat (i+1) n i.is_lt, ring_hom.comp_apply, truncated_witt_vector.coeff_truncate],
+  -- this is a bit unfortunate
+  congr' with _,
+  simp only [fin.coe_last, fin.coe_cast_le],
+end
+
+variable (f)
+
+/--
+Given compatible ring homs from `S` into `truncated_witt_vector n` for each `n`, we can lift these
+to a ring hom `S → 𝕎 R`.
+
+`lift` defines the universal property of `𝕎 R` as the inverse limit of `truncated_witt_vector n`.
+-/
+def lift : S →+* 𝕎 R :=
+by refine_struct { to_fun := lift_fun f };
+   { intros,
+     rw [← sub_eq_zero, ← ideal.mem_bot, ← infi_ker_truncate, ideal.mem_infi],
+     simp [ring_hom.mem_ker, f_compat] }
+
+variable {f}
+
+@[simp] lemma truncate_lift (s : S) :
+  witt_vector.truncate n (lift _ f_compat s) = f n s :=
+truncate_lift_fun _ f_compat s
+
+@[simp] lemma truncate_comp_lift :
+  (witt_vector.truncate n).comp (lift _ f_compat) = f n :=
+by { ext1, rw [ring_hom.comp_apply, truncate_lift] }
+
+/-- The uniqueness part of the universal property of `𝕎 R`. -/
+lemma lift_unique (g : S →+* 𝕎 R) (g_compat : ∀ k, (witt_vector.truncate k).comp g = f k) :
+  lift _ f_compat = g :=
+begin
+  ext1 x,
+  rw [← sub_eq_zero, ← ideal.mem_bot, ← infi_ker_truncate, ideal.mem_infi],
+  intro i,
+  simp only [ring_hom.mem_ker, g_compat, ←ring_hom.comp_apply,
+    truncate_comp_lift, ring_hom.map_sub, sub_self],
+end
+
+omit f_compat
+include hp
+
+/-- The universal property of `𝕎 R` as projective limit of truncated Witt vector rings. -/
+@[simps] def lift_equiv : {f : Π k, S →+* truncated_witt_vector p k R // ∀ k₁ k₂ (hk : k₁ ≤ k₂),
+  (truncated_witt_vector.truncate hk).comp (f k₂) = f k₁} ≃ (S →+* 𝕎 R) :=
+{ to_fun := λ f, lift f.1 f.2,
```

Comment:
I wonder if relying on the `mathlib` definitions of limits in src\category_theory\limits could be useful. For instance, can't this be inferred from is_limit.uniq_cone_morphism of src\category_theory\limits\limits.lean? Similarly, is_limit.hom_ext of the same file would be exactly hom_ext.

Moreover, it seems to me that introducing some category theory would allow to use creation of limits, which would also take care of my previous comment on line 378. If you can prove that the Witt vectors (as set) are a limiting cone of the set-valued functor assigning to `n` the truncated Witt vectors of length `n`, then it will also be a limiting cone for sets, groups, monoids, etc... for free since the forgetful functor creates limits. And we have this in `mathlib` (although we don't have the direct statement that forgetful functors create limit). I think this would also help in the ordering-set business.

But I might be too nearsighted and this might simply be a nightmare and many things which "would look the same to me" would not to Lean.

Reply:
This is a good point. We should definitely work on this. At the moment it's not so easy to move between the concrete world and the categorical world.
I think this is an important issue that will only become more important in the future. But I don't see how to make this painfree at the moment.

Reply:
I see, I'd also like to work on this. But I agree that for this PR what is in here is more than enough.